crates/sui-source-validation/src/lib.rs

```diff
@@ -396,3 +474,273 @@ fn local_modules(
 
     Ok(map)
 }
+
+/// Ensures `compiled_units` are compiled with the right compiler version, based on
+/// Move.lock contents. This works by detecting if a compiled unit requires a prior compiler version:
+/// - If so, download the compiler, recompile the unit, and return that unit in the result.
+/// - If not, simply keep the current compiled unit.
+fn units_for_toolchain(
+    compiled_units: &Vec<(PackageName, CompiledUnitWithSource)>,
+) -> anyhow::Result<Vec<(PackageName, CompiledUnitWithSource)>> {
+    if std::env::var("VERIFIED_SOURCE_BUILD").is_err() {
+        return Ok(compiled_units.clone());
+    }
```

Comment:
Versioned compilation is guarded by this `VERIFIED_SOURCE_BUILD` env variable right now.  I'd like this to be a way of testing the feature functionality against Move 2024 changes before Move 2024 release. But, we can also flip this so that the default is **opt in** and the env variable is **opt out**. Thoughts? I don't have big qualms about making this **opt out** and enabling toolchain versioning by default.

Note in general: `--skip-dependency-verification` exists, but this doesn't control the new changes to explicitly test regressions in version toolchain compilation (which is what this env variable allows).

Reply:
It's a bit more verbose but I'm going with `SUI_RUN_TOOLCHAIN_BUILD`, to be replaced by `SUI_IGNORE_TOOLCHAIN_BUILD` for opt-out